### PR TITLE
unify tuple expressions

### PR DESCRIPTION
**Pull request by krux02**
There is an inconsistency with how tuple expressions are represented in the AST. Sometimes they are ``nnkPar`` and sometimes they are ``nnkTupleConstr``. This makes it hard and error prone to deal tuple expressions in macros. This PR ensures that all tuple expressions are consistently of ``nnkTupleConstr``.

An issue that it solves is, if a programmer investigates the ast to look how a tuple in constructed and sees that it is done with `nnkPar`. He will do it as well. Just to realize later that this won't provide the expected result anymore if the `nnkPar` only has a single child.

After the change:

 * The parser won't generate `nnkPar` nodes anymore with more than a single child.
 * All `nnkPar` nodes that come from the parser will be those that are for operator precedence
 * All tuple construction nodes will be of kind ``nnkTupleConstr``.
 * The AST spec will have clean distinct entries for tuples and parenteses.

Example:

```nim
import macros

# covered by VarTuple (not affected here)
# let
#   (a,b) = (1,2)
#   (c,d,) = (1,2,)
#   (e,) = (1,)

macro foo(arg: untyped) =
  result = newStmtList(arg)

  for n in arg[0]:
    let expr = n[2]
    let name = n[0]
    let isTupleExpectation = newLit(expr.kind == nnkTupleConstr)
    echo expr.lispRepr
    result.add quote do:
      doAssert(`name` is tuple == `isTupleExpectation`)


  for n in arg[1]:
    let expr = n[1]
    let name = n[0]
    let isTupleExpectation = newLit(expr.kind == nnkTupleConstr)
    echo expr.lispRepr
    result.add quote do:import macros

# covered by VarTuple (not affected here)
# let
#   (a,b) = (1,2)
#   (c,d,) = (1,2,)
#   (e,) = (1,)

macro foo(arg: untyped) =
  result = newStmtList(arg)

  for n in arg[0]:
    let expr = n[2]
    let name = n[0]
    let isTupleExpectation = newLit(expr.kind == nnkTupleConstr)
    echo expr.lispRepr
    result.add quote do:
      doAssert(`name` is tuple == `isTupleExpectation`)


  for n in arg[1]:
    let expr = n[1]
    let name = n[0]
    let isTupleExpectation = newLit(expr.kind == nnkTupleConstr)
    echo expr.lispRepr
    result.add quote do:
      doAssert(`name` is tuple == `isTupleExpectation`)

  echo result.repr

foo:
  let
    a = (1,2,3)
    b = (1,2,3,)
    c = (1,)
    d = (x: 123, y: 456)
    e = (x: 789,)
    f = (y: 987)
    g = (789)  ## actually not a tuple

  var
    tmp1: (int, float)
    tmp2: (int,)
    tmp3: (int) ## actually not a tuple

      doAssert(`name` is tuple == `isTupleExpectation`)

  echo result.repr

foo:
  let
    a = (1,2,3)
    b = (1,2,3,)
    c = (1,)
    d = (x: 123, y: 456)
    e = (x: 789,)
    f = (y: 987)
    g = (789)  ## actually not a tuple

  var
    tmp1: (int, float)
    tmp2: (int,)
    tmp3: (int) ## actually not a tuple
```

### CODE REVIEW

File: tests/ast_pattern_matching.nim [hunk 1]
```diff
@@ -133,8 +133,8 @@ proc matchLengthKind*(arg: NimNode; kind: NimNodeKind; length: int): MatchingErr
   matchLengthKind(arg, {kind}, length)
 
 proc matchValue(arg: NimNode; kind: set[NimNodeKind]; value: SomeInteger): MatchingError {.compileTime.} =
-  let kindFail   = not(kind.card == 0 or arg.kind in kind)
-  let valueFail  = arg.intVal != int(value)
+  template kindFail: bool  = not(kind.card == 0 or arg.kind in kind)
+  template valueFail: bool = arg.intVal != int(value)
```
Comment (Clyybber) [inline tests/ast_pattern_matching.nim:137]: Why this change?

Comment (krux02) [reply tests/ast_pattern_matching.nim:137]: I introduced these local variables for better readability in my library ast-pattern-matching. There was a problem though with short circuit behavior in the test I introduced. That is a bug in ast-pattern-matching that I wasn't aware of. I also fixed it in that library. It is already pushed there.

**Comment by Araq** [comment]
It's a breaking change and breaks Nimble packages.

**Comment by Varriount** [comment]
Is there any way to make this backwards compatible (or at least less of a breaking change)?

As much as I like backwards compatibility, this is a rather odd case in the AST, and something that could easily be overlooked. Warts like these make macros (which are already hard to write) that much more difficult.

**Comment by juancarlospaco** [comment]
I dont think theres really too much people using 1 item tuples, so thats not a huge problem whatsoever.

`static: ` test would be nice.

`--legacy:oldTupleAst` ?

File: changelog.md [hunk 2]
```diff
@@ -63,6 +63,8 @@
   types: Type conversions cannot be used to hide `.raise` effects or side
   effects, instead a `cast` must be used. With the flag `--useVersion:1.0` the
   old behaviour is emulated.
+- Tuple expressions are now parsed consistently parsed as
```
Comment (narimiran) [inline changelog.md]: `parsed consistently parsed`

**Comment by krux02** [comment]
> I dont think theres really too much people using 1 item tuples, so thats not a huge problem whatsoever.

That is true. But I have a feeling that I don't get the message. Can you elaborate?

> ``--legacy:oldTupleAst``

Good idea. Currently the parser doesn't have the config object. But I'll see what I can do about it.

**Comment by nigredo-tori** [comment]
@krux02, even with the change you've sent to [classy](https://github.com/nigredo-tori/classy), the tests fail with this Nim branch. It looks like `NimNode` values get garbled when passing them to a macro via a `static` parameter (classy is hacky like that :frowning:). Here's a minified example:
```nim
import macros

type Foo = object
  tree: NimNode

var foo {.compileTime.}: Foo

macro save(v: untyped): untyped =
  foo = Foo(tree: v)

macro load(v: static[Foo]): untyped =
  echo "Tree from var:"
  echo treeRepr(foo.tree)
  echo "Tree from argument:"
  echo treeRepr(v.tree)
  assert: treeRepr(foo.tree) == treeRepr(v.tree)

save((x: int) => x)
load foo
```

This produces the following output:

```
Tree from var:
Infix
  Ident "=>"
  TupleConstr
    ExprColonExpr
      Ident "x"
      Ident "int"
  Ident "x"
Tree from argument:
Infix
  Ident "=>"
  TupleConstr
    Ident "int"
  Ident "x"
```
, and the assertion fails. You can see that the `ExprColonExpr` (a parameter declaration) is replaced with a type ident.

The same example seems to work OK with the current devel branch, so this seems to be an issue with this PR.

**Comment by arnetheduck** [comment]
ooh, this would be nice for `nlvm` as well, I remember it being confusing when trying to implement the codegen for tuples

**Comment by krux02** [comment]
@nigredo-tori I just start to understand the nastiness of the issue you reported. Sorry this is really a late reply. But this doesn't look like a nastiness on your side of things. This is nim internal. Why is this node changed?

**Comment by timotheecour** [comment]
I like it. These should be addressed though:
* this doesn't currently work with `()` which returns nnkPar instead of nnkTupleConstr
* regression: `(a: 1)` now printed as `(a: 1,)` via repr, it should stay `(a: 1)` which is a tuple, unlike `(a)`

File: changelog.md [hunk 3]
```diff
@@ -272,6 +272,10 @@ proc mydiv(a, b): int {.raises: [].} =
   an experimental feature. In other words, you don't have to write pragma
   `{.experimental: "forLoopMacros".}` if you want to use them.
 
+- Tuple expressions are now parsed consistently as `nnkTupleConstr` nodes.
+  This will affect macros expecting nodes to be of `nnkPar`.
+
+
```
Comment (Clyybber) [inline changelog.md:278]:
```suggestion
```

**Comment by Clyybber** [comment]
@nigredo-tori @krux02 This modification of the tuple node happens because the VM strips `nkExprColonExpr` from tuples: https://github.com/nim-lang/Nim/blob/devel/compiler/vm.nim#L2189

This simple adapted snippet:
```nim
import macros

type Foo = object
  tree: NimNode

var foo {.compileTime.}: Foo

macro save(v: untyped): untyped =
  foo = Foo(tree: v)

macro load(v: static[Foo]): untyped =
  echo "Tree from var:"
  echo treeRepr(foo.tree)
  echo "Tree from argument:"
  echo treeRepr(v.tree)
  assert: treeRepr(foo.tree) == treeRepr(v.tree)

save((x: int, ) => x) # trailing comma here makes it break
load foo
```
fails without this PR too.

**Comment by nigredo-tori** [comment]
@Clyybber, @krux02, I can confirm that with https://github.com/nigredo-tori/classy/pull/5 classy tests pass with the current version of this PR, so I went ahead and merged that one. Note that [vegansk/nimfp](https://github.com/vegansk/nimfp) still uses the old version. I'll cut both releases once this PR is merged.

**Comment by avahe-kellenberger** [comment]
Any reason this is still not appropriate for a merge @Araq? Seems like a positive change, and a solution for compatibility was mentioned

**Comment by timotheecour** [comment]
@Clyybber thanks for fixing the issues mentioned in https://github.com/nim-lang/Nim/pull/13793#issuecomment-618985932 ; 

one more thing: the tests in tests/astspec/tastspec.nim are overly complicated and weak:
eg if you change this:
```nim
  scope:
    let ast = myquote:
      (1, 2, 3)
      (a: 1, b: 2, c: 3)
      (1,)
      (a: 1)
      ()
```
to, say, 
```nim
  scope:
    let ast = myquote:
      (a: 1)
      (1, 2, 3)
      (a: 1)
      (a: 1, b: 2, c: 3)
      (1,)
      ()
```
(which adds some lines and reorders lines), the tests still pass.


Please add a file `tests/stdlib/tmacros.nim` with following content:

```nim
discard """
  targets: "c js"
"""

import std/[macros, unittest]

block: # lispRepr
  macro lispRepr2(a: untyped): string = newLit a.lispRepr

  check lispRepr2(()) == """(TupleConstr)"""
  check lispRepr2((a: 1)) == """(TupleConstr (ExprColonExpr (Ident "a") (IntLit 1)))"""
  check lispRepr2((a: 1, b: 2)) == """(TupleConstr (ExprColonExpr (Ident "a") (IntLit 1)) (ExprColonExpr (Ident "b") (IntLit 2)))"""
  check lispRepr2((1,)) == """(TupleConstr (IntLit 1))"""
  check lispRepr2((1, 2)) == """(TupleConstr (IntLit 1) (IntLit 2))"""
  check lispRepr2((1, 2, 3.0)) == """(TupleConstr (IntLit 1) (IntLit 2) (FloatLit 3.0))"""
  check lispRepr2((1)) == """(Par (IntLit 1))"""
  check lispRepr2((1+2)) == """(Par (Infix (Ident "+") (IntLit 1) (IntLit 2)))"""

block: # repr
  macro repr2(a: untyped): string = newLit a.repr

  check repr2(()) == "()"
  check repr2((a: 1)) == "(a: 1)"
  check repr2((a: 1, b: 2)) == "(a: 1, b: 2)"
  check repr2((1,)) == "(1,)"
  check repr2((1, 2)) == "(1, 2)"
  check repr2((1, 2, 3.0)) == "(1, 2, 3.0)"
  check repr2((1)) == "(1)"
  check repr2((1+2)) == "(1 + 2)"
```

which is a much simpler and more direct way to test this

File: compiler/vm.nim [hunk 4]
```diff
@@ -2185,29 +2185,36 @@ proc evalStaticStmt*(module: PSym; g: ModuleGraph; e: PNode, prc: PSym) =
...
   ## strip nkExprColonExpr from tuple values recurively. That is how
   ## they are expected to be stored in the VM.
 
   # Early abort without copy. No transformation takes place.
   if arg.kind in nkLiterals:
     return arg
 
+  if arg.kind == nkExprColonExpr and arg[0].typ != nil and
+     arg[0].typ.sym != nil and $arg[0].typ == "NimNode" and
+     arg[0].typ.owner == g.systemModule:
```
Comment (Araq) [inline compiler/vm.nim]: Is there a better solution for this gross hack? It's not acceptable.

Comment (Clyybber) [reply compiler/vm.nim]: @Araq Perhaps a new node kind that wraps around NimNode, in oder to treat static[NimNode] as a value and not AST.
I suppose we could also forbid static[NimNode], but that's a breaking change (although using static[NimNode] is not a good idea rn, regardless of this change)

File: compiler/vm.nim [hunk 5]
```diff
@@ -2185,29 +2185,36 @@ proc evalStaticStmt*(module: PSym; g: ModuleGraph; e: PNode, prc: PSym) =
 proc setupCompileTimeVar*(module: PSym; g: ModuleGraph; n: PNode) =
   discard evalConstExprAux(module, g, nil, n, emStaticStmt)
 
-proc prepareVMValue(arg: PNode): PNode =
+proc prepareVMValue(g: ModuleGraph, arg: PNode): PNode =
   ## strip nkExprColonExpr from tuple values recurively. That is how
```
Comment (timotheecour) [inline compiler/vm.nim]: pre-existing but s/recurively/recursively

File: compiler/vm.nim (hunk 4 above)
Comment (timotheecour) [reply compiler/vm.nim]: this should work:
```nim
  if arg.kind == nkExprColonExpr and arg[0].typ != nil and arg[0].typ.sym != nil and
    arg[0].typ.sym.magic == mPNimrodNode:
```

or simpler, with `import std/wrapnils` now that it doesn't depend on dotoperators:
```nim
if arg.kind == nkExprColonExpr and ?.arg[0].typ.sym.magic == mPNimrodNode:
```

That said, can this branch ever be exercised? which code generates a `static NimNode`? eg:
```nim
when true:
  import macros
  macro bar1(a: static int) = discard
  macro bar2(a: static NimNode) = discard
  macro baz() =
    let a = newLit 0
    bar1(1) # ok
    bar2(newLit 0) # Error: type mismatch: got <NimNode>
    discard
  baz()
```
(note: if the goal was to support genericity in macros, this is made possible by https://github.com/nim-lang/Nim/pull/11992 which supports `macro fn[T](a: static T)` )

Comment (Clyybber) [reply compiler/vm.nim]: Applied timothees suggestion, should be slightly less of a hack now (it still is one, but it's there to support the static NimNode hack :)

Comment (timotheecour) [reply compiler/vm.nim]: but when can a static NimNode occur? at least add a comment mentioning it's not clear if this branch can ever be taken

Comment (timotheecour) [reply compiler/vm.nim]: also @Clyybber your commit (which you then reverted) used:
```
 if arg.kind == nkExprColonExpr and arg[0].typ != nil and
     arg[0].typ.sym != nil and arg[0].sym.magic == mPNimrodNode:
```

instead of
```
  if arg.kind == nkExprColonExpr and arg[0].typ != nil and arg[0].typ.sym != nil and
    arg[0].typ.sym.magic == mPNimrodNode:
```
which i had suggested above and should work

Comment (Clyybber) [reply compiler/vm.nim]: oh thanks, nice catch

File: changelog.md [hunk 6]
```diff
@@ -272,6 +272,8 @@ proc mydiv(a, b): int {.raises: [].} =
   an experimental feature. In other words, you don't have to write pragma
   `{.experimental: "forLoopMacros".}` if you want to use them.
 
+- Tuple expressions are now parsed consistently as `nnkTupleConstr` nodes.
```
Comment (timotheecour) [inline changelog.md]: changes like this require either bumping NimPatch (and implementing `nimVersionCT`, by re-opening/re-visiting https://github.com/nim-lang/Nim/pull/14648) or adding this to condsyms.nim:
```nim
defineSymbol("nimHasUnifiedTuple")
```

so that code can be written with fwd and bwd compatibility when needed

File: changelog.md [hunk 7]
```diff
@@ -248,6 +248,171 @@
 
 - Added `hasClosure` to `std/typetraits`.
 
+# 1.2 - xxxx-xx-xx
```
Comment (timotheecour) [inline changelog.md]: rebase problem

File: compiler/vm.nim (hunk 5 above)
Comment (timotheecour) [reply compiler/vm.nim]: it's ok to leave this as future work but then please don't mark this as resolved when it's not otherwise this is forgotten

Comment (timotheecour) [reply compiler/vm.nim]: ah i see 1a407402a4614fc9100617f71c2d3c9b8a8367a5, thanks! adding a comment there would've helped clear this up, but ok

Comment (xflywind) [reply compiler/vm.nim]: maybe make a small PR to clean them
https://github.com/nim-lang/Nim/commit/1a407402a4614fc9100617f71c2d3c9b8a8367a5

File: changelog.md [hunk 8]
```diff
@@ -314,6 +314,9 @@
 
 - `--hint:CC` now goes to stderr (like all other hints) instead of stdout.
 
+- Tuple expressions are now parsed consistently as
```
Comment (timotheecour) [inline changelog.md]: how about move this to `Language changes` ? in future work we should clarify in the template `changelogs/changelog_X_XX_X.md` what belongs in each category (with examples)

Comment (Clyybber) [reply changelog.md]: Yeah, I thought about moving it to `Language changes` too; if you agree then I'll do it.

File: tests/macros/tdumpast.nim [hunk 9]
```diff
@@ -48,3 +48,9 @@ macro fun3(): untyped =
     int | float | array | seq | object | ptr | pointer | float32
   doAssert n.repr == "int | float | array | seq | object | ptr | pointer | float32", n.repr
 fun3()
+
```
Comment (timotheecour) [inline tests/macros/tdumpast.nim:51]: https://github.com/nim-lang/Nim/pull/13793#issuecomment-792524597 still holds

Comment (Clyybber) [reply tests/macros/tdumpast.nim:51]: Added to `tdumpast.nim` instead, as this is more related to the compiler than the stdlib.

File: doc/astspec.txt [hunk 10]
```diff
@@ -329,19 +329,55 @@ AST:
...
+Tuple Constructors
+------------------
+
+Nodes for tuple construction are built with the ``nnkTupleConstr`` node.
+
+Concrete syntax:
+
+.. code-block:: nim
+  (1, 2, 3)
+  (a: 1, b: 2, c: 3)
```
Comment (timotheecour) [inline doc/astspec.txt:356]:
```suggestion
  (a: 1, b: 2, c: 3)
  ()
```

(plus corresponding AST below)

File: tests/ast_pattern_matching.nim (hunk 1 above)
Comment (timotheecour) [reply tests/ast_pattern_matching.nim:137]: looks like we can revert these 2 lines then?

Comment (timotheecour) [inline tests/ast_pattern_matching.nim:137]:
```suggestion
  let kindFail = not(kind.card == 0 or arg.kind in kind)
  let valueFail = arg.intVal != int(value)
```

File: changelog.md [hunk 11]
```diff
@@ -272,6 +272,9 @@
 
 - Custom numeric literals (e.g. `-128'bignum`) are now supported.
 
+- Tuple expressions are now parsed consistently as
+  `nnkTupleConstr` node. Will affect macros expecting nodes to be of `nnkPar`.
```
Comment (timotheecour) [inline changelog.md:276]: can you list here a representative set of cases that are changed from nnkPar to/from nnkTupleConstr ?

File: tests/ast_pattern_matching.nim (hunk 1 above)
Comment (Clyybber) [reply tests/ast_pattern_matching.nim:137]: I don't think that's what @krux02 meant.

File: changelog.md (hunk 11 above)
Comment (Clyybber) [reply changelog.md:276]: I think that would be more confusing, since the old logic wasn't really consistent.

Comment (Clyybber) [reply changelog.md:276]: An exhaustive list would be more of a documentation of the old inconsistent behaviour, rather than the new one, so I'm wary of doing that.

**Review by timotheecour — APPROVED** [review]
excellent!

**Comment by ringabout** [comment]
> It's a breaking change and breaks Nimble packages.

Indeed, it breaks weave https://github.com/mratsim/weave/issues/173

I will try to fix it.
